File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        tf = defaultdict(int)
        for ch in s1:
            tf[ch] += 1
        sf = defaultdict(int)
        l  = 0
        def check(tf, sf):
            for k, v in tf.items():
                if k not in sf or sf[k] != v:
                    return False
            return True
        for r in range(len(s2)):
            sf[s2[r]] += 1
            print(sf)
            if r >= len(s1):
                sf[s2[l]] -= 1
                if not sf[s2[l]]:
                    del sf[s2[l]]
                l += 1
            if check(tf, sf):
                return True

        return False

        

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    private boolean check(int []a, int[] b) {
        for (int i = 0; i < a.length; i++) {
            if (a[i] > 0) {
                if (b[i] < a[i]) return false;
            }
        }
        return true;
    }
    public boolean checkInclusion(String s1, String s2) {
        int freq2[] = new int[26];
        int freq1[] = new int[26];
        for (char ch: s1.toCharArray()) {
            freq1[ch-'a']++;
        }
        int l = 0, n1 = s1.length(), n2 = s2.length();

        for (int r = 0; r < n2; r++) {
            freq2[s2.charAt(r)-'a']++;
            if (r - l + 1 > n1) {
                freq2[s2.charAt(l++) - 'a']--;
            }
            if (check(freq1, freq2)) {
                return true;
            }
        }

        return false;
    }
}